feat: dynamic theming based on sway vars and GTK

boredland opened the pull request:
Resolves https://github.com/Manjaro-Sway/manjaro-sway/issues/99

History: Initially I wanted to achieve dynamic theming by using `pywal` and `wpgtk`, but in my experience these tools are overly complex and the resulting color scheme lacks readability, so, the branch name is kind of misleading :D

I've been working on this PR for the last couple of days, and this is a big revamp in how themes are configured.
Main changes:
1. sed and linking are not used anymore, all configs within userspace are editable and belong to the user
2. All apps that have an ability to setup colors via CLI are set up using colors from sway theme's definitions file.
3. All apps that don't have such ability are using GTK theme's `@define-color` tags.

I've liked the result more than `pywal` approach because it's simpler to customize and predictable enough not to cause bugs afterwards.

Basically that means that if you will change your GTK theme and write a custom definitions file for that theme, every app from the set of defaults (except foot, don't have any ideas on how to configure it for now. Maybe creating sequences file for zsh alike to `pywal`, but I don't like file tricks in dotfiles)

Current issues:
1. Somehow the modes indication is missing from `waybar` now. I don't know how I broke it. Didn't investigate much, help would be appreciated. Modes themselves work fine
2. I've deleted Kvantum theming, mainly because it was the only thing relying on templating, and as far as I understand, Manjaro's common dotfiles already include qt5ct and Kvantum configs. I don't know if these themes can be set programmatically (like with GTK and `gsettings`). I don't know what those themes are responsible for, either.
3. Abeforementioned foot issue, for now it's not themed in any way except the default config in `~/.config/foot/foot.ini`, which is set to (near) `matcha-dark-sea` colors.
4. Subjectively, using GTK colors makes `waybar` default text color a bit dull. I could use text-selected property from Matcha theme, but that would make text `#ffffff`, which doesn't look great either

Thorough testing of this PR is required, because my dotfiles are kind of a mess at this point, bug reports, feedback and new ideas are welcome, of course.

FallenChromium committed:
Rename colors to match pywal template

FallenChromium committed:
Delete matcha-blue theme, maintenance burden

FallenChromium committed:
Edit sway modes to use new color template

FallenChromium committed:
Delete kitty template, kitty isn't used anymore

FallenChromium committed:
Remove unnecessary gtk settings.ini changes

FallenChromium committed:
Disable Qt theme changes for now (manjaro has their default)

FallenChromium committed:
Remove unnecessary gtk templates (can be done via gsettings)

FallenChromium committed:
Remove unnecessary mako template, can be done via sway vars

FallenChromium committed:
Remove unnecessary swaylock config, can be done via sway vars

FallenChromium committed:
Remove unnecessary .css theme (waybar|wlogout), can be done via gtk theme vars

FallenChromium committed:
Refactor matcha-green theme definitions

FallenChromium committed:
Use GTK theme colors in Waybar

FallenChromium committed:
Use GTK theme colors in wlogout

FallenChromium committed:
Remove script for mako theming, now done via CLI arguments

FallenChromium committed:
Delete script for linking .css themes, GTK theme is used instead

FallenChromium committed:
Fix wob accent color

FallenChromium committed:
Add variable for notification daemon. Now using CLI args for theming

FallenChromium committed:
Start $notifications instead of static mako command

FallenChromium committed:
Merge branch 'sway' into pywal-theming

FallenChromium committed:
Fix mako dynamic theming

Apparently the previous command didn't work for several reasons
1. --margin 20% was incorrect, it should be --margin 20
2. Font and colors were used incorrectly (only first argument was read, sway makes no assumptions and thus doesn't sanitize vars by default), and mako is intolerant to such arguments

FallenChromium committed:
Make rofi font dynamic

Was set statically before. With Default $gui-font value ('Roboto 11') it looks a bit small, if there are no reasons against, it would probably make sense to change it to 'Roboto 14' for consistency.

Mathew-D commented:
![rofi with old](https://user-images.githubusercontent.com/44036272/126741921-a65bb875-030a-4a57-88fe-5f5182b18436.png)
-the help icon on waybar will not run ** FIXXED **
-when the screen cap section is on you can't open mod + d
-I can't get rofi to run.  I get this error 
![rofi error](https://user-images.githubusercontent.com/44036272/126733310-a50d5ced-9e12-46f7-ad11-290fb03b762a.png)
** I fixed the error but the theme is missing the background color on the items can you look ** Just for reference before I switched it to theme after adding the # I got the color above which is wrong.   If we want we could find a good theme online and just paste it to one line with the vars as the colors.

Mathew-D commented:
- removed my mako config, and the colors didn't work from default theme **Fixed now**
- on shutdown, I see failed to parse option 'text-color'.  I think it is coming from gset **the include fixed this I think**
- waybar's colors are ok but the text color is a little gray instead of more white.
- wlogout colors are correct.

Mathew-D committed:
fix help on waybar

The " was in the wrong spot and it broke the help.

Mathew-D committed:
Rofi theme

To use the theme vars you have to include them before.  
You were missing the # in front of each of the colors in your method of changing colors and I was getting the wrong colors.
I think I changed it to the new theming method as per: https://github.com/davatorium/rofi/blob/next/doc/rofi.1.markdown#layout-and-theming

Mathew-D committed:
Mako theming

Now that the theme is included mako almost worked.  It still needed # in front of the colors.
Also the notify stayed forever.  We should make it time out.

Mathew-D commented:
As the theme has to be included before we can use it we might want to make a new file for user theming like: 
 ~/.config/sway/user_theme
And inside do something like:
#If you wish to select your own theme please uncomment the line below and give the path the the folder that has the definition file in it.
#set $theme ~/where_theme_is_definition_is_located

Then in /etc/sway/definitions  we can go something like this:
[ -z "$theme" ] && set $theme /usr/share/sway/themes/matcha-green

Mathew-D commented:
I didn't try the lock but I think it might need some # on the colors like the rest did.

FallenChromium commented:
Damn, I'm sorry about the colors, I forgot to add a commit that changes the accents defintions to the format workout alpha and with a hex sign, I did that locally but not in the repo. I will push it, but then some changes that you made should be reversed. Nice catch with the help menu (I disabled it in my waybar so I couldn't check that) and with the mako timeout (though most apps, like Telegram Desktop pass a custom timeout option anyway)

Mathew-D commented:
Ok, push your changes and remove my commits and I will take a look.

FallenChromium committed:
Fix additional variables in stock theme definitions file

Mathew-D committed:
undo changes to rofi and mako for color

undo some changes after change to theme file.

Mathew-D committed:
add include

include your theme variables

Mathew-D committed:
move theme var

Moved theme include out of definitions

Mathew-D committed:
moved theme include

Moved theme include to main config.

Mathew-D committed:
Java fix

fix java programs.

Mathew-D committed:
done in another commit

done in another pull

FallenChromium committed:
Add support for wob >= 0.22

FallenChromium committed:
Merge branch 'dynamic-theming' of https://github.com/Manjaro-Sway/desktop-settings into dynamic-theming

boredland commented on community/sway/etc/sway/definitions:
```diff
@@ -60,11 +58,10 @@ set $shutdown wlogout --buttons-per-row 3
 # onscreen bar
-set $onscreen_bar bash /usr/share/sway/scripts/wob.sh
-
+set $onscreen_bar bash /usr/share/sway/scripts/wob.sh "$accent-colorFF" "$background-colorFF" 
```
does sway escape that correctly? i'd have assumed that it searches for a variable `accent-colorFF`. interesting this works

FallenChromium replied:
It does, and I learned it the hard way, while using $term-font, it gave me `foot-font` and not `Roboto NerdFonts`. :|

FallenChromium replied:
It does match for as much as possible while within quotes though.

boredland replied:
irritating, but ok.

boredland commented on community/sway/usr/share/sway/scripts/wob.sh:
```diff
@@ -12,17 +13,15 @@ is_running_on_this_screen() {
 }
 
-[[ "$new_value" ]] && echo $new_value > $wob_pipe
+[[ "$new_value" ]] && echo $new_value > $wob_pipe     
```
```suggestion
[[ "$new_value" ]] && echo $new_value > $wob_pipe
```

FallenChromium replied:
Sorry :D

boredland commented on community/sway/usr/share/sway/scripts/wob.sh:
```diff
@@ -12,17 +13,15 @@ is_running_on_this_screen() {
 # wob does not appear in $(swaymsg -t get_msg), so:
 is_running_on_this_screen wob || {
-    tail -f $wob_pipe | wob --border-color "#FF${color}" --bar-color "#FF${color}" --background-color '#FF000000' --anchor top --anchor center --margin 20 &
+    tail -f $wob_pipe | wob --border-color $1 --bar-color $1 --background-color $2 --anchor top --anchor center --margin 20 &
```
```suggestion
    tail -f $wob_pipe | wob --border-color "$1" --bar-color "$1" --background-color "$2" --anchor top --anchor center --margin 20 &
```

boredland replied:
otherwise "#" would comment stuff out

FallenChromium replied:
I believe it doesn't work like that in the bash scripts, the variables (arguments in this case) should get escaped. Definitely works for me without quotes

boredland replied:
ya, I tried and it seems to work nicely. nvm

FallenChromium committed:
Delete trailing spaces

Co-authored-by: Jonas Strassel <info@jonas-strassel.de>
